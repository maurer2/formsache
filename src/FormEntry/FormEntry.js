import React from 'react';
import PropTypes from 'prop-types';
import styles from './FormEntry.module.scss';

const FormEntry = ({label, name, value, type, handleChange, isValid})=> (
  <div className={`${styles.formEntry} ${isValid ? '' : styles.formEntryIsInvalid} `}>
    <label className={styles.formLabel} htmlFor={name}>
      {label}
    </label>
    <input
      className={styles.formField}
      value={value}
      onChange={handleChange}
      type={type}
      name={name}
      id={name}
    />
  </div>
);

export default FormEntry;

FormEntry.propTypes = {
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  isValid: PropTypes.bool.isRequired,
};
