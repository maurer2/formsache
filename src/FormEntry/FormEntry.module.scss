.formEntry {
  display: flex;
  margin-bottom: 2rem;
  flex-direction: column;
}

.formLabel {
  display: flex;
  margin-bottom: 0.75rem;
}

.formField {
  padding: 0.75rem 1rem;
  width: 100%;
  min-width: 0;
  border: 2px solid black;
  outline-offset: -1px;
}

.formEntryIsInvalid {
  color: #de313f;

  .formLabel:after {
    content: '\00d7';
    margin-left: auto;
    font-size: 1.15rem;
    font-weight: bold;
    line-height: 1;
  }

  .formField {
    border-color: red;
  }
}