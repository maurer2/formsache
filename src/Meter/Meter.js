import React from 'react';
import style from './Meter.module.scss';

const Meter = ({ min, max, value }) => {
  const translations = {
    1: 'bad',
    2: 'middle',
    3: 'good',
  };

  const showDescription = Object.prototype.hasOwnProperty.call(translations, value);

  return (
    <div className={style.wrapper}>
      <meter className={style.meter} min={min} max={max} value={value} />
      { showDescription && (
        <dl className={style.description}>
          <dt className={style.key}>Strength: </dt>
          <dd className={style.value}>{translations[value]}</dd>
        </dl>
      )}
    </div>
  );
};

export default Meter;
