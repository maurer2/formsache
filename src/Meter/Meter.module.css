.wrapper {
  border: 1px solid black;
}

.meter {
  margin: auto;
  width: 50%;
  height: 1rem;
}

.meter::-webkit-meter-bar {
  background: none;
  background-color: none;
  background-image: linear-gradient(
    to right, 
    red 0%,
    red 32.667%,
    transparent 32.667%,
    transparent 33.667%,
    red 33%,
    red 65.337%,
    transparent 65.337%,
    transparent 66.337%,
    red 66.337%,
    red 100%
  );
  background-size: 100% 100%;
}
