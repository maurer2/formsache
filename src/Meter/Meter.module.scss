@function calculate-gradient($total, $color-bar, $color-spacing) {
  $gradients: ();
  $length: percentage(1 / $total);

  @for $i from 1 to ($total + 1) {
    $is-not-last-iteration: ($i != $total);
    $start: ($i - 1) * $length;
    $end: $i * $length;

    $bar: $color-bar $start, $color-bar calc(#{$end} - #{$step-spacing});
    $spacing: $color-spacing calc(#{$end} - #{$step-spacing}), $color-spacing $end;

    $gradient: $bar;

    // last entry doesn't have spacing right
    @if $is-not-last-iteration {
      $gradient: join($bar, $spacing);
    }

    $gradients: join($gradients, $gradient, comma);
  }

  @return $gradients;
}

$meter-background: #fff;
$step-background: lightgray;
$step-foreground: red;
$step-spacing: 10px;

.wrapper {
  border: 1px solid black;
}

.meter {
  position: relative;
  margin: auto;
  width: 50%;
  height: 1rem;
  background: $meter-background;
  background: $step-background;

  // gutter overlay
  &:after {
    $gradients: calculate-gradient(3, transparent, $meter-background);

    content: '';
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background-image: linear-gradient(to right, $gradients);
    z-index: 55;
  }
}

// background
.meter::-webkit-meter-bar {
  $gradients: calculate-gradient(3, $step-background, $meter-background);

  background: none;
  background-color: inherit;
  background-image: linear-gradient(to right, $gradients);
  background-size: 100%;
  background-repeat: no-repeat;
  box-shadow: none;
}

// actual-bar
.meter::-moz-meter-bar {
  background: none;
  background-color: $step-foreground;
  box-shadow: none;
}

// actual bar
.meter::-webkit-meter-optimum-value,
.meter::-webkit-meter-optimum-value,
.meter::-webkit-meter-suboptimum-value,
.meter::-webkit-meter-even-less-good-value {
  background: none;
  background-color: $step-foreground;
}

.description {
  display: flex;
}

.key {
  font-weight: bold;
}

.value {
  margin: 0;
}
