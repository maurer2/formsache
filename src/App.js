import React, { useState, useEffect } from 'react';
import { default as passwortStrengthChecker } from 'zxcvbn';

import Meter from './Meter/Meter';
import FormEntry from './FormEntry/FormEntry';
import Overlay from './Overlay/Overlay';

import styles from './App.module.scss';

function App() {
  // field values
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  // validity flags
  const [emailIsValid, setEmailIsValid] = useState(true);
  const [passwordIsValid, setPasswordIsValid] = useState(true);
  const [passwordStrength, setPasswordStrength] = useState(0);
  // submit flag
  const [formIsSubmittable, setFormIsSubmittable] = useState(false);
  // overlay flag
  const [showOverlay, setShowOverlay] = useState(false);

  // watch submittability of form
  useEffect(() => {
    const emailIsFilled = email.length > 0;
    const passwordIsFilled = password.length > 0;

    const formIsReady = emailIsFilled && emailIsValid && passwordIsValid && passwordIsFilled;

    setFormIsSubmittable(formIsReady);
  }, [ email, password ]);

  const handleEmailChange = (event) => {
    event.preventDefault();

    const { value } = event.target;
    setEmail(value);

    const emailIsNotEmpty = email.length > 0;
    // https://emailregex.com/
    const emailRegex = new RegExp(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
    const isValidEmail = emailRegex.test(value);

    setEmailIsValid(isValidEmail && emailIsNotEmpty);
  }

  const handlePasswordChange = (event) => {
    event.preventDefault();

    const { value } = event.target;
    setPassword(value);

    const passwordIsNotEmpty = password.length > 0;
    const { score } = passwortStrengthChecker(value);
    // contains a number
    const passwordRegex = new RegExp(/\d/);
    const isValidPasword = passwordRegex.test(value) && score >= 2;

    setPasswordStrength(score);
    setPasswordIsValid(isValidPasword && passwordIsNotEmpty);
  }

  const handleSubmit = (event) => {
    event.preventDefault();

    if (!formIsSubmittable) {
      return;
    }
    setShowOverlay(true);
  };

  return (
    <form className={styles.form} onSubmit={handleSubmit} action="#" autoComplete="off">
      <fieldset className={styles.formRow}>
        <FormEntry
          label="E-Mail"
          name="email"
          type="text"
          placeholder="test@test.de"
          value={email}
          handleChange={handleEmailChange}
          isValid={emailIsValid}
        />
        <FormEntry
          label="Password"
          name="password"
          type="password"
          placeholder="Must contain a number"
          value={password}
          handleChange={handlePasswordChange}
          isValid={passwordIsValid}
        />
      </fieldset>
      <div className={styles.formRow}>
        <Meter value={passwordStrength} min={0} max={3}/>
      </div>
      <button
        className={`${styles.formButton} ${!formIsSubmittable ? styles.formButtonIsDisabled : ''}`}
        disabled={!formIsSubmittable}
        type="submit"
      >
        Submit
      </button>
      { showOverlay && <Overlay message="Login successful!" /> }
    </form>
  );
}

export default App;
