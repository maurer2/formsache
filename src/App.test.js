import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

import Meter from './Meter/Meter';
import FormEntry from './FormEntry/FormEntry';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  test('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('child components exist', () => {
    expect(wrapper.find('.form').exists()).toBe(true);
    expect(wrapper.find('.formRow').exists()).toBe(true);
    expect(wrapper.find(FormEntry).exists()).toBe(true);
    expect(wrapper.find('[name="email"]').exists()).toBe(true);
    expect(wrapper.find('[name="password"]').exists()).toBe(true);
    expect(wrapper.find(Meter).exists()).toBe(true);
    expect(wrapper.find('.formButton').exists()).toBe(true);
  });
});
