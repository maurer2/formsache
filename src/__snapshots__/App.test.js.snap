// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<form
  action="#"
  autoComplete="off"
  className="form"
  onSubmit={[Function]}
>
  <fieldset
    className="formRow"
  >
    <FormEntry
      handleChange={[Function]}
      isValid={true}
      label="E-Mail"
      name="email"
      placeholder="test@test.de"
      type="text"
      value=""
    />
    <FormEntry
      handleChange={[Function]}
      isValid={true}
      label="Password"
      name="password"
      placeholder="Must contain a number"
      type="password"
      value=""
    />
  </fieldset>
  <div
    className="formRow"
  >
    <Meter
      max={3}
      min={0}
      value={0}
    />
  </div>
  <button
    className="formButton formButtonIsDisabled"
    disabled={true}
    type="submit"
  >
    Submit
  </button>
</form>
`;
